atom/browser/api/gpuinfo_manager.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2018 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "atom/browser/api/gpuinfo_manager.h"
+#include "atom/browser/api/gpu_info_enumerator.h"
+#include "base/memory/singleton.h"
+#include "base/threading/thread_task_runner_handle.h"
+#include "content/public/browser/browser_thread.h"
+#include "gpu/config/gpu_info_collector.h"
+
+namespace atom {
+
+GPUInfoManager* GPUInfoManager::GetInstance() {
+  return base::Singleton<GPUInfoManager>::get();
+}
+
+GPUInfoManager::GPUInfoManager()
+    : gpu_data_manager_(content::GpuDataManagerImpl::GetInstance()) {
+  gpu_data_manager_->AddObserver(this);
+}
+
+GPUInfoManager::~GPUInfoManager() {
+  content::GpuDataManagerImpl::GetInstance()->RemoveObserver(this);
+}
+
+// Based on
+// https://chromium.googlesource.com/chromium/src.git/+/66.0.3359.181/content/browser/gpu/gpu_data_manager_impl_private.cc#810
+bool GPUInfoManager::NeedsCompleteGpuInfoCollection() {
+  const auto& gpu_info = gpu_data_manager_->GetGPUInfo();
+#if defined(OS_MACOSX)
+  return gpu_info.gl_vendor.empty();
+#elif defined(OS_WIN)
+  return (gpu_info.dx_diagnostics.values.empty() &&
+          gpu_info.dx_diagnostics.children.empty());
+#else
+  return false;
+#endif
+}
+
+// Should be posted to the task runner
+void GPUInfoManager::ProcessCompleteInfo() {
+  const auto result = EnumerateGPUInfo(gpu_data_manager_->GetGPUInfo());
+  // We have recieved the complete information, resolve all promises that
+  // were waiting for this info.
+  for (const auto& promise : complete_info_promise_set_) {
+    promise->Resolve(*result);
+  }
+}
+
+void GPUInfoManager::OnGpuInfoUpdate() {
+  // Ignore if called when not asked for complete GPUInfo
+  if (NeedsCompleteGpuInfoCollection())
+    return;
+  base::ThreadTaskRunnerHandle::Get()->PostTask(
+      FROM_HERE, base::BindOnce(&GPUInfoManager::ProcessCompleteInfo,
+                                base::Unretained(this)));
+}
+
+// Should be posted to the task runner
+void GPUInfoManager::CompleteInfoFetcher(util::Promise* promise) {
+  complete_info_promise_set_.insert(promise);
+
+  if (NeedsCompleteGpuInfoCollection()) {
+    gpu_data_manager_->RequestCompleteGpuInfoIfNeeded();
+  } else {
+    GPUInfoManager::OnGpuInfoUpdate();
+  }
+}
+
+void GPUInfoManager::FetchCompleteInfo(util::Promise* promise) {
+  base::ThreadTaskRunnerHandle::Get()->PostTask(
```

Comment:
Why does this post to the task runner ? The result from `OnGpuInfoUpdate` seems to be already made available on the next tick.

Reply:
Yes, but `OnGpuInfoUpdate` will not be called if complete request has been fetched already once. So, if `NeedsCompleteGpuInfoCollection` is false, then the call needs to be run on the next tick. I can change it to post only from the else condition in `GPUInfoManager::CompleteInfoFetcher` is you feel that's better.